{caddy2,h2o}: use hash of path as ETag in Nix store

emilazy opened the pull request:
###### Motivation for this change

These match the behaviour of the existing nginx patch.

It would be nice to not have to patch every web server under the sun, but unfortunately I don't see a better option; ETag calculation is pretty uniformly this naive. A potential alternative, though wider-reaching than is in scope for a simple nixpkgs PR like this one, would be to set the mtime of files in the Nix store to an arbitrary date based on their store hash. (There's some precedent for this in other systems; I forget the details, but one of the fields in PKI used to be a date but is now filled with some arbitrary entropy.)

@GrahamcOfBorg build caddy2 h2o

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

emilazy commented:
Rebased to fix conflict; cc @thoughtpolice?

yegortimoshenko commented:
cc @aszlig @7c6f434c

yegortimoshenko commented on pkgs/servers/http/h2o/nix-store-etag.patch:
```diff
@@ -0,0 +1,63 @@
++
+     if (ref->_etag.len == 0)
+         ref->_etag.len = sprintf(ref->_etag.buf, "\"%08x-%zx\"", (unsigned)ref->st.st_mtime, (size_t)ref->st.st_size);
++
```
Nitpick: probably doesn't need to be patched.

emilazy replied:
No longer relevant, since we're patching these lines anyway.

aszlig replied:
As far as I can see we're not touching `outbuf` nor are we modifying the `memcpy`, did I miss something?

emilazy replied:
`h2o_filecache_in_nix_store`, defined below, calls this function with `outbuf = NULL` to check whether a cache entry is in the Nix store. It's possible this could be factored better, but all the ways I could think of doing it were more invasive, and especially we want to benefit from the caching of the E-Tag in `lib/handler/file.c:do_send_file` rather than checking multiple times.

aszlig reviewed (changes requested):
Thanks a lot for your contribution :-)

While the things I commented are pretty minor, I'm however strongly against merging this without *any* tests whatsoever.

Can you please add a NixOS VM test for these web servers as well?

You could probably generalise our [WebDriver test for nginx](https://github.com/NixOS/nixpkgs/blob/df21a1e977ad200a32f99a3871cfe6917c45a662/nixos/tests/nginx-etag.nix) to work for caddy2 and h2o as well.

emilazy commented on pkgs/servers/http/h2o/nix-store-etag.patch:
```diff
@@ -0,0 +1,63 @@
+ 
+-#define H2O_FILECACHE_ETAG_MAXLEN (sizeof("\"deadbeef-deadbeefdeadbeef\"") - 1)
++/* + 2 to account for the surrounding quotes */
++#define H2O_FILECACHE_ETAG_MAXLEN (@nixStoreHashLen@ + 2)
```
It's the maximum length, not the buffer size, [which has 1 added to it](https://github.com/h2o/h2o/blob/b4775b5320ad5061761bd72d5a8092d49e2e4f93/include/h2o/filecache.h#L47):

```c
            struct {
                char buf[H2O_FILECACHE_ETAG_MAXLEN + 1];
                size_t len;
            } _etag;
`

emilazy commented on pkgs/servers/http/h2o/nix-store-etag.patch:
```diff
@@ -0,0 +1,63 @@
++        char *respath = realpath(ref->_path, NULL);
++        if (respath != NULL && strncmp(respath, nix_store_prefix, nix_store_prefix_len) == 0) {
++            /* set etag to store path hash */
++            ref->_etag.len = sprintf(ref->_etag.buf, "\"%.@nixStoreHashLen@s\"", respath + nix_store_prefix_len);
```
It will never overflow: the precision specifier `%.Ns` writes at most N bytes from the source, so the maximum number of bytes written is `@nixStoreHashLen@ + sizeof("\"\"")` = `@nixStoreHashLen@ + (2 + 1)` = `H2O_FILECACHE_ETAG_MAXLEN + 1` = `sizeof(ref->_etag.buf)`.

I'd probably use `snprintf` if it was my own code, but the existing code uses `sprintf` so I matched it for consistency.

aszlig replied:
Ah, thanks for elaborating on that, given that the allocated size of `buf` is `H2O_FILECACHE_ETAG_MAXLEN + 1` this makes a lot more sense.

emilazy commented on pkgs/servers/http/h2o/default.nix:
```diff
@@ -14,6 +14,15 @@ stdenv.mkDerivation rec {
+      src = ./nix-store-etag.patch;
+      nixStoreDir = builtins.storeDir;
+      nixStoreHashLen = with stdenv.lib;
+        stringLength (head (builtins.split "-" (baseNameOf "${src}")));
```
I didn't realize this counted as an import-from-derivation :(

I'd like to avoid it if possible; I'm happy to just hard-code `32`, but would prefer a nicer way to get the store hash length in a future-proof manner. Unfortunately, I can't think of one...

aszlig replied:
What about `strchr` towards `-` at runtime like we do in the nginx patch?

yegortimoshenko replied:
Doesn't look like IFD to me. Isn't this pretty much the same as `stdenv.mkDerivation { /* ... */ buildPhase = "${coreutils}/bin/ls"; }`?

yegortimoshenko replied:
> [...] so it won't build in read-only mode.

Hm, did you mean that it won't evaluate? `nix-instantiate --readonly-mode --eval -A h2o` works, nothing will build in read-only mode by definition.

aszlig replied:
@yegortimoshenko: Correct, I meant eval here.

And yes, @emilazy also: I was wrong in regards to IFD: Since the store path doesn't actually *need* to exist for evaluation, it isn't an IFD (also, there is no `import` involved, duh!). Sorry for the noise.

Infinisil replied:
@aszlig Note that IFD is a bit of a misnomer, because the same effect can happen with a `builtins.readFile <derivation>` (notably without an `import`). A better name for IFD would be "derivation-dependent evaluation"

emilazy commented:
Tests would be nice, though it would be a little awkward without NixOS modules, and I'm not sure there are tests for things that don't have integration at all like that? I'd like to add a module with test for at least one of these servers, but that's blocked on deciding which to switch to :)

For what it's worth, I've tested both locally with various conditions (symlinks in and out of the store, etc.) and they match the behaviour of the nginx path (though more efficient: the nginx path causes at least one unconditional syscall and allocation for every request, which I think is unfortunate). I'll see if I can find the time to generalize the existing nginx test and fix the conflict.

aszlig commented:
> Tests would be nice, though it would be a little awkward without NixOS modules

Ah, didn't check whether NixOS modules actually exist. Doing a quick & dirty `systemd.services` definition in the tests until we have modules should suffice I guess, it's more like a "don't forget about tests for the etag patch" for future NixOS module authors.

lukateras committed:
nixos/tests/common/etag: split out of nixos/tests/nginx-etag

lukateras committed:
nixos/tests/h2o-etag: init

lukateras committed:
nixos/tests/caddy2-etag: init

emilazy commented:
@aszlig You were right to demand tests; these were broken due to failing to clear Last-Modified :)

Updated with fixes, refactoring, and tests (thanks to @yegortimoshenko). I also switched to using `snprintf` in the H2O patch, just in case `sizeof(ref->_etag.buf)` changes -- it's not part of the patch context so there's no airtight guarantee of safety in the face of unpredictable h2o changes without that. diffs are a weird format to work with.

Infinisil commented on lib/trivial.nix:
```diff
@@ -332,4 +332,34 @@ rec {
+
+     Currently 32, but you never know!
+  */
+  storeHashLength = builtins.stringLength fakeStoreHash;
```
The definitions for `fakeDerivation`, `fakeStorePath` and `fakeStoreHash` are only used by this definition here, and are arguably not of much use for other parts of the code base. So I suggest to just defined them as a `let in` under `storeHashLength`.

emilazy replied:
Okay, made them local for now.

@GrahamcOfBorg test caddy-etag h2o-etag

emilazy commented:
@GrahamcOfBorg test caddy-etag h2o-etag

emilazy committed:
caddy2: use hash of path as ETag in Nix store

This is more correct than the behaviour of the nginx/H2O patches in some
edge cases (documented in the patch) while having negligible overhead
(the cost of resolving symlinks in the path far outweighs the cost of
the hash function; remember that HTTPS involves a lot of cryptographic
operations to begin with).

Co-authored-by: Yegor Timoshenko <yegortimoshenko@riseup.net>

emilazy commented:
Updated to optimize the Caddy patch to avoid syscalls unless necessary (guarding `filepath.Abs` and `filepath.EvalSymlinks`), as well as increase correctness by addressing the corner case mentioned in https://github.com/NixOS/nixpkgs/pull/76697#issuecomment-569687912.

Microbenchmark with [`h2load(1)`](https://nghttp2.org/documentation/h2load.1.html) running `caddy file-server ... -root /nix/store/cksvzjk11hhkshhfd3y9xbgjnpi3zvhl-root`:

```
nixpkgs:       finished in 4.53s, 22053.46 req/s, 26.97MB/s
initial patch: finished in 4.85s, 20615.56 req/s, 25.19MB/s
current patch: finished in 4.50s, 22209.55 req/s, 27.14MB/s
```

As you can see, I have managed to increase the performance of a web server in the course of making it perform more operations, and am therefore a wizard.

@GrahamcOfBorg test h2o-etag caddy2-etag

aszlig commented:
@emilazy: Very cool :-) Will do a full review later, hopefully before Thursday.

Side note about performance: We could also avoid the call to `realpath()` (which potentially involves several syscalls) for the nginx patch and *only* really do it if we know that `last_modified_time == 1`.

emilazy committed:
h2o: use hash of path as ETag in Nix store

This matches the behaviour of the caddy2 patch;
see 3e8b3c7bd00c275688fe0096e0c6ea080b18205f.

emilazy commented:
Well, I underestimate my own masochism: H2O might not have a real language but it does have OpenSSL to hand! Updated to match the Caddy patch, and I'm going to look into adjusting the nginx patch too. This also means the `storeHashLength` definition can be dropped. Benchmarks:

```
nixpkgs:       198503.90 req/s, 243.83MB/s
initial patch: 192906.03 req/s, 236.96MB/s
current patch: 198045.72 req/s, 243.08MB/s
```

H2O is really fast! Fun fact: OpenSSL doesn't let you do SHA512/224 or BLAKE2b hashing without either a heap allocation for each digest or piping around a silly `EVP_MD_CTX *` everywhere (and that that type includes, um, [a vtable](https://github.com/openssl/openssl/blob/OpenSSL_1_1_1/crypto/evp/evp_locl.h#L12-L22)). I just used the much simpler but technically deprecated SHA-512 interface; I don't find [the upstream advice](https://www.openssl.org/docs/man1.1.1/man3/EVP_DigestInit.html#NOTES) terribly convincing:

> The **EVP** interface to message digests should almost always be used in preference to the low level interfaces. This is because the code then becomes transparent to the digest used and much more flexible.

especially when it shares a manual page with:

> <dl>
> <dt>EVP_MD_pkey_type()</dt>
> <dd>Returns the NID of the public key signing algorithm associated with this digest. For example EVP_sha1() is associated with RSA so this will return <b>NID_sha1WithRSAEncryption</b>. Since digests and signature algorithms are no longer linked this function is only retained for compatibility reasons.</dd>
> </dl>

@GrahamcOfBorg test caddy2-etag h2o-etag

emilazy commented:
Updated to fix an embarrassing NUL-termination bug in the H2O patch, confirming my suspicions about my ability to write correct C code. (The outputted headers seemed fine, but I can't tell you whether that's due to callers guarding with `.len`, or just luck.)

I also decided against backporting these changes to nginx; I started moving the Nix store ETag calculation code to `ngx_http_static_module.c` to more properly fix the issue that #80671 worked around, but ran into the issue that it'd need patching `ngx_http_{gzip,brotli}_static_module.c` too and likely involve a lot of code duplication. So I gave up instead.

If anyone does want to adjust the nginx patch, my suggested approach to allow `ngx_http_set_etag` to set the ETag correctly would be to have `ngx_open_cached_file` cache the relevant path information in `ngx_open_file_info_t`, and store a pointer to the latter in `ngx_http_request_t`.  It's kind of a mess, though.

@GrahamcOfBorg test caddy2-etag h2o-etag

emilazy commented:
Rebased. cc @aszlig again, if you don't mind taking a look :)

@GrahamcOfBorg test nginx-etag caddy2-etag h2o-etag

emilazy commented:
Rebased.

@ofborg test h2o-etag caddy2-etag

emilazy commented:
cc @aszlig again, if you don't mind taking a look? ^^

emilazy commented:
Rebased again to fix a dangling commit hash reference in a commit message and slightly refactor `h2o_filecache_in_nix_store`.

@ofborg test caddy2-etag h2o-etag

emilazy commented on pkgs/servers/http/h2o/nix-store-etag.patch:
```diff
@@ -0,0 +1,137 @@
++         * it's measurable!) */
++        char *respath = NULL;
++        if (strncmp(ref->_path, nix_store_prefix, nix_store_prefix_len) == 0 ||
++            ((respath = realpath(respath, NULL)) != NULL &&
```
Oops. Yes, you're right; I guess the test is working because it has a direct `/nix/store` prefix rather than going through a symlink, so it benefits from the happy-path optimization. I'll correct the patch and extend the test to check this case tomorrow.

emilazy commented:
Hm, for some reason my reply to https://github.com/NixOS/nixpkgs/pull/83111#discussion_r429723274 isn't showing up here, so I'll post it inline:

> `h2o_filecache_in_nix_store`, defined below, calls this function with `outbuf = NULL` to check whether a cache entry is in the Nix store. It's possible this could be factored better, but all the ways I could think of doing it were more invasive, and especially we want to benefit from the caching of the E-Tag in `lib/handler/file.c:do_send_file` rather than checking multiple times.

Mic92 commented:
@aszlig anything else required?

aszlig commented:
@Mic92: Yep, see the [unresolved comment](https://github.com/NixOS/nixpkgs/pull/83111#discussion_r429721663) above.

omasanori commented:
ping?
